Enable periodic scans.

tomprince opened the pull request:
Fixes #138.

hacklschorsch commented on src/magic_folder/config.py:
```diff
@@ -1443,9 +1464,31 @@ def create_magic_folder(self, name, magic_path, author,
+                ),
+            )
+        if scan_interval is not None and (
+            not isinstance(scan_interval, int) or scan_interval <= 0
```
[In the DB above, 0 is an allowed value](https://github.com/LeastAuthority/magic-folder/pull/468/files#diff-db422d0a72330e38baed0b613e6f69f9f154b6e67e0db758516627d0492fb634R147), but here 0 is forbidden.

tomprince replied:
I've added a constraint to the DB.

hacklschorsch commented on src/magic_folder/config.py:
```diff
@@ -1443,9 +1464,31 @@ def create_magic_folder(self, name, magic_path, author,
+        ):
+            raise APIError(
+                code=http.BAD_REQUEST,
+                reason="Scan interval must be a positive integer or null (not '{!r}').".format(
```
`null` or `0` ? Now my confusion is complete 😸

tomprince replied:
`null`

hacklschorsch commented on src/magic_folder/migrate.py:
```diff
@@ -81,7 +81,8 @@ def magic_folder_migrate(config_dir, listen_endpoint_str, tahoe_node_directory,
             mf_config[u'upload_dircap'],
-            mf_config[u'poll_interval'],  # is this always available?
+            int(mf_config[u'poll_interval']),  # is this always available?
+            int(mf_config[u'poll_interval']),
```
should this maybe be `scan_interval` ?

tomprince replied:
tahoe-lafs didn't have that (which is where this is migrating from. I've added a comment to clarify.

hacklschorsch commented on src/magic_folder/service.py:
```diff
@@ -221,6 +224,13 @@ def create_folder(self, name, author_name, local_dir, poll_interval):
+        if scan_interval is not None and scan_interval < 0:
+            raise APIError(
+                code=http.BAD_REQUEST,
+                reason="scan_interval must be positive integer or null",
```
...or `0`?

hacklschorsch commented on src/magic_folder/test/cli/test_magic_folder.py:
```diff
@@ -592,6 +652,61 @@ def test_poll_interval(self):
+                "add",
+                "--name", "test",
+                "--author", "test",
+                "--poll-interval=-0",
```
And what happens with plain `0` here?  I expect a `UsageError` to be thrown?

tomprince replied:
I think this was a bad copy-paste from the `-1` (where I changed the `1` and didn't remove the `0`. Fixed.

hacklschorsch commented on src/magic_folder/test/cli/test_magic_folder.py:
```diff
@@ -592,6 +652,61 @@ def test_poll_interval(self):
+                "add",
+                "--name", "test",
+                "--author", "test",
+                "--scan-interval=-0",
```
And what happens with plain `0` here?  I expect it to work and to turn off local scanning.

hacklschorsch commented on src/magic_folder/test/test_web.py:
```diff
@@ -365,6 +366,7 @@ def magic_folder_config(author, local_directory):
         u"collective-dircap": u"URI:DIR2-RO:{}:{}".format(b2a("\0" * 16), b2a("\1" * 32)),
         u"upload-dircap": u"URI:DIR2:{}:{}".format(b2a("\2" * 16), b2a("\3" * 32)),
         u"poll-interval": 60,
+        u"scan-interval": None,
```
Is `None` different from `null` and `0`?

meejah replied:
`null` is JS terminology and is basically the same as `None` (i.e. a `null` in some JSON will become a `None` on the python side and vice-versa). Python doesn't have "`null`".

hacklschorsch replied:
Thanks meejah!

My comment was about our using the three terms seemingly interchangeably.  I was a bit uncompromising maybe with having the word `null` used in docs meaning the number `0`. Users will probably figure that one out.

From my reading the patch I believe currently sometimes `None` and sometimes `0` means _do not scan at all_.
It should be one of those two; I personally like `0` here better since it's of the same type as the other valid values.

meejah replied:
Yeah, definitely "choose one" I'd say.
I also prefer `0` to mean "disable scanning" (which is also somewhat common for these sorts of options IME).

tomprince replied:
I think the only place where `0` was mentioned was in the `--help` output (and it isn't supported. I've removed that mention of it. `0` shouldn't be supported, and `null` (in the API/db)/ `None` in python means no scanning.

hacklschorsch commented on src/magic_folder/test/test_web.py:
```diff
@@ -522,6 +526,36 @@ def test_add_folder_invalid_json(self):
+                        loads,
+                        MatchesDict(
+                            {
+                                "reason": StartsWith("scan_interval must be positive integer or null"),
```
So I begin to believe `null` equals `0`, while `None` equals `not set`. Correct?

If this is so in the rest of the application, `null` as the word (as opposed to `positive integer`) is probably better English than `0` here... the digit `0` might be less confusing still

hacklschorsch reviewed (commented):
Sweet!

I found one typo between `scan` and `poll` that IMHO should be fixed; the rest is musings about the (in)equality between `0`, `null` and `None` and how `scan_interval` seems (to me) to not be fully consistent regarding this.

meejah commented:
@tomprince I think my notifications say "requested review" .. do you mean on top of the review @hacklschorsch already provided? Also CI is red currently too (so either Flo or I can take another pass after the changes Flo suggests and CI is green).

tomprince commented on src/magic_folder/config.py:
```diff
@@ -1443,9 +1464,31 @@ def create_magic_folder(self, name, magic_path, author,
+            (in seconds).
+
+        :param Optional[int] scan_interval: how often to scan for local changes
+            (in seconds).
```
No, I intended for `0` to be disallowed, and `None` to indicate turning it off. I've fixed the help in the cli usage.

tomprince commented:
> @tomprince I think my notifications say "requested review" .. do you mean on top of the review @hacklschorsch already provided? Also CI is red currently too (so either Flo or I can take another pass after the changes Flo suggests and CI is green + conflicts resolved).

I requested that before the review @hacklschorsch did. In any case, I think I've addressed the issues.

hacklschorsch commented:
I believe @tomprince requested a review by @meejah, but I did one anyway - thought I might be able to help out :) Seems I added confusion. Sorry. Anyway, thanks for addressing my points @tomprince ; Great work as always!

meejah commented:
okay, if things are fixed per Flo's review once conflicts are resolved and CI are green please merge.
